java-checks/src/main/java/org/sonar/java/checks/PrivateFieldUsedLocallyCheck.java

```diff
@@ -0,0 +1,285 @@
+      for (CFG.Block block : liveFields.cfg.reversedBlocks()) {
+        liveFields.fieldsReadOnAnotherInstance.addAll(fieldsReadOnAnotherInstance(block));
+        Set<Symbol> blockKill = new HashSet<>();
+        Set<Symbol> blockGen = new HashSet<>();
+        liveFields.processBlockElements(block, blockKill, blockGen);
+        kill.put(block, blockKill);
+        gen.put(block, blockGen);
+      }
+      liveFields.analyzeCFG(in, kill, gen);
+      // out of exit block are empty by definition.
+      if (!liveFields.out.get(liveFields.cfg.reversedBlocks().get(0)).isEmpty()) {
+        throw new IllegalStateException("Out of exit block should be empty");
+      }
+
+      liveFields.in = in;
+      return liveFields;
+    }
+
+    private static Set<Symbol> fieldsReadOnAnotherInstance(Block block) {
+      Set<Symbol> fieldsReadOnAnotherInstance = new HashSet<>();
+      for (Tree element : block.elements()) {
+        if (element.is(Kind.MEMBER_SELECT)) {
+          MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) element;
+          if (memberSelect.expression().is(Kind.IDENTIFIER)) {
+            String objectName = ((IdentifierTree) memberSelect.expression()).name();
+            Symbol symbol = memberSelect.identifier().symbol();
+
+            if (!symbol.isStatic() && !"this".equals(objectName) && isField(symbol)) {
+              fieldsReadOnAnotherInstance.add(symbol);
+            }
+          }
+        }
+      }
+      return fieldsReadOnAnotherInstance;
+    }
+
+    private void analyzeCFG(Map<CFG.Block, Set<Symbol>> in, Map<CFG.Block, Set<Symbol>> kill, Map<CFG.Block, Set<Symbol>> gen) {
+      Deque<Block> workList = new LinkedList<>();
+      workList.addAll(cfg.reversedBlocks());
+      while (!workList.isEmpty()) {
+        CFG.Block block = workList.removeFirst();
+
+        Set<Symbol> blockOut = out.get(block);
+        if (blockOut == null) {
+          blockOut = new HashSet<>();
+          out.put(block, blockOut);
+        }
+        for (CFG.Block successor : block.successors()) {
+          Set<Symbol> inOfSuccessor = in.get(successor);
+          if (inOfSuccessor != null) {
+            blockOut.addAll(inOfSuccessor);
+          }
+        }
+        // in = gen and (out - kill)
+        Set<Symbol> newIn = new HashSet<>(gen.get(block));
+        newIn.addAll(Sets.difference(blockOut, kill.get(block)));
+
+        if (newIn.equals(in.get(block))) {
+          continue;
+        }
+        in.put(block, newIn);
+        for (CFG.Block predecessor : block.predecessors()) {
+          workList.addLast(predecessor);
+        }
+      }
+    }
+
+    private void processBlockElements(CFG.Block block, Set<Symbol> blockKill, Set<Symbol> blockGen) {
+      // process elements from bottom to top
+      Set<Tree> assignmentLHS = new HashSet<>();
+      for (Tree element : Lists.reverse(block.elements())) {
+        Symbol symbol;
+        switch (element.kind()) {
+          case ASSIGNMENT:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1151)